### PR TITLE
Simplified callback method notifications.

### PR DESCRIPTION
**Pull request by sbordet**
Instead of using allocating lambdas, now the callback methods:

* onSuccess()
* onFailure()
* onAborted()
* onCompleted()

are notified directly, and exceptions thrown by those methods are logged and ignored.

Also reviewed other corner cases, especially in the handling of abort() and close(). Simplified and strengthened the state machine, and added test cases.

### CODE REVIEW

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/IteratingCallback.java [hunk 1]
```diff
@@ -522,18 +526,64 @@ else if (onAbortedOnFailureIfNotPendingDoCompleted != null)
...
+    private void notifyFailure(Throwable failure)
+    {
+        try
+        {
+            onFailure(failure);
+        }
+        catch (Throwable x)
+        {
+            if (LOG.isDebugEnabled())
+                LOG.debug("onFailure() failure", x);
```
Comment (gregw) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/IteratingCallback.java:552]: This should add suppressed to keep the same behaviour as before
```suggestion
            addSuppressedIfNotAssociated(failure, x);
            if (LOG.isDebugEnabled())
                LOG.debug("onFailure() failure", x);
```

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/IteratingCallback.java [hunk 2]
```diff
@@ -522,18 +526,64 @@ else if (onAbortedOnFailureIfNotPendingDoCompleted != null)
...
+
+    private void notifyAborted(Throwable failure)
+    {
+        try
+        {
+            onAborted(failure);
+        }
+        catch (Throwable x)
+        {
+            if (LOG.isDebugEnabled())
```
Comment (gregw) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/IteratingCallback.java:565]: add suppressed

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/IteratingCallback.java [hunk 3]
```diff
@@ -522,18 +526,64 @@ else if (onAbortedOnFailureIfNotPendingDoCompleted != null)
...
+
+    private void notifyCompleted(Throwable failure)
+    {
+        try
+        {
+            onCompleted(failure);
+        }
+        catch (Throwable x)
+        {
+            if (LOG.isDebugEnabled())
```
Comment (gregw) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/IteratingCallback.java:579]: add suppressed

**Review by gregw — CHANGES_REQUESTED** [review]

**Review by lorban — CHANGES_REQUESTED** [review]
Besides adding `addSuppressedIfNotAssociated()` like @gregw already reported, LGTM.

**Review by lorban — APPROVED** [review]
